
## Summary

1. List of which similarity blocks have been computed and an estimate for when the entire set might be completed
   - [ ] Revised size estimate for the N x N matrix and output size
   - [ ] Revised computational effort (cup hours) estimate for N x N matrix

2. Summary statistics on the metagenome assemblies for each block
   - [ ] Number of reads input
   - [ ] Number of contigs output and N50
   - [ ] Number of proteins called from each caller
   - [ ] Number of unique proteins in the block
   - [ ] Mean number of internally similar proteins at X cutoff (distribution)
   - [ ] Size distribution and GC ratios
   - [ ] Fraction with confident function assignments
   - [ ] Fraction with no similar hits in sample and DB36
   - [ ] Estimate of number of novel proteins

3. Contiguity and pathway related estimates
   - [ ] Fraction of proteins on a contig with > x orfs (2, ..., n)
   - [ ] Fraction estimated to be novel operons (all novel orfs)
   - [ ] Fraction estimated to linked to known pathways

4. Rates of discovery
   - [ ] Rate of novel proteins per megabase of assembled contigs (MoAC)
   - [ ] Rate of novel operons per MoAC
   - [ ] Expansion rate of known protein family members per MoAC
   - [ ] Global estimate of novel proteins
   - [ ] Global estimate of unique proteins

## Explanations for Estiamtes

### 1. List of which similarity blocks have been computed and an estimate for when the entire set might be completed
#### Revised size estimate for the N x N matrix and output size
#### Revised computational effort (cup hours) estimate for N x N matrix

### 2. Summary statistics on the metagenome assemblies for each block
#### Number of reads input
#### Number of contigs output and N50
#### Number of proteins called from each caller
#### Number of unique proteins in the block
#### Mean number of internally similar proteins at X cutoff (distribution)
#### Size distribution and GC ratios
#### Fraction with confident function assignments
#### Fraction with no similar hits in sample and DB36
#### Estimate of number of novel proteins

### 3. Contiguity and pathway related estimates
#### Fraction of proteins on a contig with > x orfs (2, ..., n)
#### Fraction estimated to be novel operons (all novel orfs)
#### Fraction estimated to linked to known pathways

### 4. Rates of discovery
#### Rate of novel proteins per megabase of assembled contigs (MoAC)
#### Rate of novel operons per MoAC
#### Expansion rate of known protein family members per MoAC
#### Global estimate of novel proteins
#### Global estimate of unique proteins
